<font size=5><b>Profit Maximization through the Implementation of Simplex Method in Home Industry for Snacks Production</b></font>

# <h3><b> A. GROUP MEMBERS

In [ ]:
# @title **This project has been authored by:**


table = [["Kristian Surya Dinata", "S2043845"],
         ["Tey Yi Jie", "17081752"],
         ["Ani Yeap", "17218658"],
         ["Chan Jie Min", "S2141167"]]

print(tabulate(table, headers=["Prepared by", "Matric No #"], tablefmt="grid"))

+-----------------------+---------------+
| Prepared by           | Matric No #   |
+=======================+===============+
| Kristian Surya Dinata | S2043845      |
+-----------------------+---------------+
| Tey Yi Jie            | 17081752      |
+-----------------------+---------------+
| Ani Yeap              | 17218658      |
+-----------------------+---------------+
| Chan Jie Min          | S2141167      |
+-----------------------+---------------+


# <h3><b> B. PROJECT OVERVIEW

<h4><b>A. Introduction</h3></b>

<p>In today's highly competitive business landscape, companies strive to achieve profitability and maintain a competitive edge. To achieve profitability, organizations must optimize their production processes and resource allocation strategies. One effective approach is to implement mathematical optimization techniques, such as linear programming, to maximize profits while considering various constraints.</p>

<p>Production activities are a complex network of operations that transform inputs into finished goods. The efficiency and effectiveness of these activities directly impact a company's ability to generate profits. To ensure profitability, companies need to carefully manage their resources, including raw materials, machinery, and labor, to achieve the desired output levels with minimal costs.</p>

<p>The simplex method is a widely used mathematical optimization technique that provides a systematic approach to solving linear programming problems. It aims to identify the optimal solution among a set of feasible solutions, considering both profit maximization and resource utilization. The simplex method does this by iteratively evaluating extreme points, or vertices, of the feasible region. By doing so, it efficiently determines the best allocation of resources, leading to an optimal production plan.</p>

<p>The snack production industry can greatly benefit from the application of the simplex method. With a wide range of snacks and various input factors involved, including ingredients, production machinery, and labor, finding the optimal combination is crucial for maximizing profits. By utilizing the simplex method, snack producers can make informed decisions on resource allocation, production volume, and product mix, ultimately leading to improved profitability.<p>

<p>In this assignment, we focus on the application of the simplex method in snacks production with the objective of profit maximization. Specifically, we examine the factors affecting profit. By formulating a linear programming model and employing the simplex method, we aim to identify the optimal resource allocation strategy that maximizes profits while satisfying various production constraints.<p>

<br>

<h4><b>B. Problem Statement</h3></b>
<p>In the home snacks production industry, it struggles to maximize their profits due to inefficient resource allocation and limited understanding of mathematical optimization techniques. The absence of a systematic approach, such as linear programming with the simplex method, hinders the ability to optimize resource utilization and achieve maximum profitability. Without proper allocation of raw materials, machinery, and labor, the industry fails to exploit the full production potential, leading to suboptimal profits. Therefore, there is a critical need to address this problem by implementing the simplex method to optimize resource allocation and maximize profits in snacks production.</p><br>

<b><h4>C. Objectives</h3></b>
<ol>
  <li>To develop and implement a linear programming model using the simplex method to optimize resource allocation in the home snacks production industry</li>
  <li>To Maximize profitability by determining the best combination of raw materials, machinery, and labour through the application of the simplex method</li>
  <li>To Evaluate the effectiveness of the optimization approach on production and profit within the snacks production process</li>
</ol>

# <h3><b> C. Literature Review

The home snacks production industry plays a vital role in meeting the growing demand for convenient and ready-to-eat snacks. To ensure long-term sustainability and competitiveness, snack producers must focus on maximizing profits through efficient resource allocation and optimization techniques. This literature review aims to explore relevant studies and research that have investigated the application of the simplex method for profit maximization in the home snacks production industry.

The simplex method, a widely recognized mathematical optimization technique, has been extensively applied to various industries for profit maximization. The simplex method provides a systematic approach to solve linear programming problems by iteratively evaluating extreme points of the feasible region. It has proven successful in optimizing resource allocation and achieving maximum profitability in manufacturing, supply chain management, and transportation sectors.

To achieve maximum profitability, snack producers need to consider various factors that affect their operations. Raw material availability, production capacity, market demand, and production costs are crucial variables that influence profit margins. Furthermore, effective utilization of machinery and labour resources is critical for profit maximization in the home snacks production industry. Optimizing the production schedule and workforce allocation using the simplex method can lead to reduced idle time, improved productivity, and ultimately, higher profits.

Several studies have evaluated the effectiveness of optimization approaches, including the simplex method, in the context of profit maximization in the snacks production industry.

In article by Uthayageetha (2017), the methodology section briefly explains the mathematical modelling of linear programming problems using decision variables and constraints. The objective is to maximize the return on different production processes. The article includes cost data for cultivating tomato and chilli crops, including various aspects such as seed cost, labour cost, fertilizers, pest control and harvesting time. The study found out that



by putting the focus on the cultivation of few main crop able to maximize their profits in agriculture.

In another study by Barbieri et al. (2019) the study addresses the decision-making process of packaging section in a Brazilian e-commerce company. They implement the study by using a linear programming model using the simplex method to maximize the company's revenue based on historical data of activities for each type of product. They determine which products should be prioritized to generate the highest revenue. The results of their analysis indicate that prioritizing certain products can lead to a higher revenue.

In a study to maximize the Bintang Bakery home industry who is facing a similar challenge, including limited raw materials, machinery operating hours, and labour hours. The lack of linear programming implementation and mathematical understanding in the production process is identified as a cause for not achieving maximum profits. To address these issues, the authors propose using the simplex method for solving linear programming problems, to optimize the profit level of the Bintang Bakery home industry (Anggoro et al., 2019).

The literature review reveals that the implementation of the simplex method in the home snacks production industry holds promise for profit maximization. The simplex method offers a systematic approach to optimize resource allocation, production schedules, and workforce utilization. By incorporating numerous factors affecting profitability into a linear programming model, snack producers can make informed decisions and achieve higher profit margins.

However, it is worth noting that further research is needed to explore specific challenges and constraints unique to the home snacks production industry. Additionally, case studies and real-world implementations are necessary to validate the effectiveness of the simplex method and its impact on profitability.

# <h3><b> D. RESULT AND DISCUSSION

## <h3> 1. Snack Production </h3>

The home industry specializes in producing three types of snacks: Cheese Puffs, Potato Chips, and Pretzels.

In [ ]:
# @title <h5> Table 1. Each Snack Detail </h5>


# Import the raw data
from pandas.io.formats.style import Styler
import pandas as pd
data = pd.read_excel('/content/numericaloptimization/DATA/product_data.xlsx')

# Create a Styler object for the DataFrame
styler = data.style

# Apply styles to the DataFrame
styler.set_table_styles([
    {'selector': 'th',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightblue'),
         ('font-weight', 'bold'),
         ('border', '1px solid black')
     ]},
    {'selector': 'td',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightgray'),
         ('border', '1px solid black')
     ]}
])

# Display the styled DataFrame
display(styler)

The Snack Factory strives to create high-quality snacks that cater to different tastes and preferences. Each snack type offers a unique flavor and texture to provide a delightful snacking experience for consumers. The company focuses on optimizing production processes and resource allocation to achieve maximum profitability while maintaining product quality and customer satisfaction.
<br><br>
The steps in the production process for the snacks (Cheese Puffs, Potato Chips, and Pretzels) are as follows:
<br><br>
**1.	Preparation of Raw Materials:**

*   Cheese Puffs: Weigh and measure the required quantity of flour, cheese, salt, vegetable oil, and spices.
*   Potato Chips: Clean, peel, and slice the potatoes into thin, uniform slices. Measure the required quantity of salt and vegetable oil.
*   Pretzels: Prepare the dough by measuring and mixing flour, salt, yeast, and water.

**2.	Mixing Ingredients / Making the Dough:**

*   Cheese Puffs: Combine the flour, cheese, salt, vegetable oil, and spices in a mixing bowl. Mix the ingredients until a dough-like consistency is achieved.
*   Potato Chips: Toss the potato slices with salt and vegetable oil to evenly coat them.
*   Pretzels: Knead the dough mixture until it becomes smooth and elastic.

**3.	Distribution of Snack Base:**
*   Cheese Puffs: Shape the dough into small, bite-sized balls and place them on a baking tray.
Potato Chips: Arrange the sliced potatoes in a single layer on baking sheets.
*   Pretzels: Divide the dough into smaller portions and shape them into pretzel twists or desired shapes.

**4.	Baking/Grilling:**
*   Cheese Puffs: Bake the cheese puff balls in a preheated oven until they turn golden brown and crispy.
*   Potato Chips: Fry the potato slices in hot oil until they become golden and crispy.
*   Pretzels: Boil the pretzels in water with baking soda, then bake them in the oven until they are golden and slightly crispy on the outside.

**5.	Snack Packaging:**
*   Cheese Puffs: Allow the cheese puffs to cool, then package them in individual List item packs or containers.
*   Potato Chips: Allow the potato chips to cool and drain excess oil, then package them in bags or containers.
*   Pretzels: Let the pretzels cool, then package them in bags or containers.

<br>
<h5>The Snack Factory follows these production steps to ensure the quality and consistency of its snacks. Each step is carefully executed to deliver tasty, well-packaged snacks to customers.

## <h3> 2. Resource Availability </h3>

The table provided below illustrates the availability of essential resources for producing snacks within a span of one month. These resources, known as production factors, play a vital role in the snack manufacturing process.

In [ ]:
# @title <h5> Table 2. Essential Resources Availability</h5>

from pandas.io.formats.style import Styler
import pandas as pd
availability = pd.read_excel('/content/numericaloptimization/DATA/availability.xlsx')

# Create a Styler object for the DataFrame
styler = availability.style

# Apply styles to the DataFrame
styler.set_table_styles([
    {'selector': 'th',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightblue'),
         ('font-weight', 'bold'),
         ('border', '1px solid black')
     ]},
    {'selector': 'td',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightgray'),
         ('border', '1px solid black')
     ]}
])

# Display the styled DataFrame
display(styler)

## <h3> 3. Resource Utilization </h3>

The table provided below illustrates the utilization of essential resources for producing snacks.

In [ ]:
# @title <h5> Table 3. Essential Resources Utilization</h5>


# Import the raw data
from pandas.io.formats.style import Styler
import pandas as pd
production = pd.read_excel('/content/numericaloptimization/DATA/production.xlsx')

# Create a Styler object for the DataFrame
styler = production.style

# Apply styles to the DataFrame
styler.set_table_styles([
    {'selector': 'th',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightblue'),
         ('font-weight', 'bold'),
         ('border', '1px solid black')
     ]},
    {'selector': 'td',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightgray'),
         ('border', '1px solid black')
     ]}
])

# Display the styled DataFrame
display(styler)

## <h3> 4. Simplex Method </h3>

To conclude the simplex method and derive an optimization model for maximizing profits in the home industry, the steps are outlined below:

### **<h5>a. Set Decision Variables**</h5>

In solving linear program problem for the home industry, the decision variables are defined to represent the different types of snacks produced. The quantities produced for each snack type are provided in the table below, with 𝑥1 representing Cheese Puffs (800 packs), 𝑥2 representing Potato Chips (365 packs), and 𝑥3 representing Pretzels (250 packs).

In [ ]:
# @title <h5>Table 4. Decision Variables </h5>

# Import the raw data
from pandas.io.formats.style import Styler
import pandas as pd
variable = pd.read_excel('/content/numericaloptimization/DATA/variable.xlsx')

# Create a Styler object for the DataFrame
styler = variable.style

# Apply styles to the DataFrame
styler.set_table_styles([
    {'selector': 'th',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightblue'),
         ('font-weight', 'bold'),
         ('border', '1px solid black')
     ]},
    {'selector': 'td',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightgray'),
         ('border', '1px solid black')
     ]}
])

# Display the styled DataFrame
display(styler)

### **<h5>b. Set Constraints**</h5>

By setting clear constraints, we simplify the production process and work towards achieving our snack production goals with greater focus and efficiency. The given constraints represent various limitations for the home industry. Each constraint is defined by an equation or inequality that should be satisfied by the decision variables 𝑥1, 𝑥2, and 𝑥3. The corresponding limits or thresholds are specified in the table above.

In [ ]:
# @title <h5>Table 5. Constraints </h5>

# Import the raw data
from pandas.io.formats.style import Styler
import pandas as pd
constraint = pd.read_excel('/content/numericaloptimization/DATA/constraint.xlsx')

# Create a Styler object for the DataFrame
styler = constraint.style

# Apply styles to the DataFrame
styler.set_table_styles([
    {'selector': 'th',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightblue'),
         ('font-weight', 'bold'),
         ('border', '1px solid black')
     ]},
    {'selector': 'td',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightgray'),
         ('border', '1px solid black')
     ]}
])

# Display the styled DataFrame
display(styler)

### **<h5>c. Set Objective Function**</h5>

Defining an objective function is a key step in snack production. The objective function outlines the specific goal or outcome we aim to maximize the production process. It serves as a crucial tool in optimizing snack production, ensuring that our actions are purposeful and directed towards achieving our production objectives.

In [ ]:
# @title <h5>Table 6. Objective Function

# Import the raw data
from pandas.io.formats.style import Styler
import pandas as pd
objective = pd.read_excel('/content/numericaloptimization/DATA/objective.xlsx')

# Create a Styler object for the DataFrame
styler = objective.style

# Apply styles to the DataFrame
styler.set_table_styles([
    {'selector': 'th',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightblue'),
         ('font-weight', 'bold'),
         ('border', '1px solid black')
     ]},
    {'selector': 'td',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightgray'),
         ('border', '1px solid black')
     ]}
])

# Display the styled DataFrame
display(styler)

The objective function of the problem is represented by the equation Z = 2500𝑥1 + 6000𝑥2 + 5000𝑥3. The coefficients of the decision variables 𝑥1, 𝑥2, and 𝑥3 indicate the contribution of each variable to the objective function. The goal is to maximize the value of Z, which represents the overall profit or optimization objective for home industry.

### **<h5>d. Inequalities Transformation**</h5>

To transform an inequality (≤) into an equation (=), we can introduce slack variables and artificial variables. The slack variables represent the surplus or slackness in meeting the constraint, while the artificial variables are introduced to convert inequality (≥) into an equation. By adding these variables to the left side of the inequality, we modify the constraint to an equality. This process allows us to handle constraints more flexibly and solve linear programming problems effectively.




In [ ]:
# @title <h5>Table 7. Inequalities Transformation

# Import the raw data
from pandas.io.formats.style import Styler
import pandas as pd
converted = pd.read_excel('/content/numericaloptimization/DATA/converted.xlsx')

# Create a Styler object for the DataFrame
styler = converted.style

# Apply styles to the DataFrame
styler.set_table_styles([
    {'selector': 'th',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightblue'),
         ('font-weight', 'bold'),
         ('border', '1px solid black')
     ]},
    {'selector': 'td',
     'props': [
         ('text-align', 'center'),
         ('background-color', 'lightgray'),
         ('border', '1px solid black')
     ]}
])

# Display the styled DataFrame
display(styler)

**Objective Function:** <br><br>
 𝑍 = 2500𝑥1 + 6000𝑥2 + 5000𝑥3 + 0𝑆1 + 0𝑆2 + 0𝑆3 + 0𝑆4 + 0𝑆5 + 0𝑆6 + 0𝑆7 + 0𝑆8 + 0𝑆9 + 0𝑆10 + 0𝑆11 + 0𝑆12 + 0𝑆13 + 0𝑆14 + 0𝑆15 + 0𝑆16 + 0𝑆17 + 0𝑆18 − 𝑚𝑎16 − 𝑚𝑎17 − 𝑚𝑎18
<br><br>
In order to convert the inequality constraints (≤) into equality constraints (=), the slack variables (𝑆1 to 𝑆18) and artificial variables (𝑎16 to 𝑎18) are introduced. The table above presents the modified constraints with the added variables. The objective function 𝑍 remains the same, including the coefficients of the decision variables (𝑥1, 𝑥2, 𝑥3) as well as the newly introduced slack and artificial variables. The artificial variables are multiplied by a large negative constant (𝑚) to discourage their inclusion in the optimal solution.


### **<h5>e. Create Simplex Table**</h5>

The simplex table organizes relevant information about the variables, constraints, and objective function. It provides a clear representation of the optimization problem at hand, facilitating the application of the simplex method for finding the optimal solution. By organizing the variables, constraints, and objective function within a simplex table, we gain insights into the problem structure and iteratively perform simplex algorithm operations to navigate towards the optimal solution

### **<h5>f. Iteration**</h5>

Each iteration involves evaluating the current solution, identifying the most favorable variable to enter or leave the solution, and updating the tableau or simplex table accordingly. Through a series of iterations, the simplex method gradually improves the objective function value until an optimal solution is reached.

In [ ]:
# @title <h5>Optimization Finding</h5>
from pulp import *
import pandas as pd
from tabulate import tabulate

# Create a maximization problem
prob = LpProblem("Snacks_Profit_Maximization", LpMaximize)

# Define the decision variables
x1 = LpVariable("x1", lowBound=800, cat='Continuous')
x2 = LpVariable("x2", lowBound=350, cat='Continuous')
x3 = LpVariable("x3", lowBound=250, cat='Continuous')

# Define the objective function
prob += 5 * x1 + 7 * x2 + 6 * x3

# Define the constraints
prob += 150 * x1 + 200 * x2 + 250 * x3 <= 350000
prob += 50 * x1 + 5 * x2 + 7 * x3 <= 250000
prob += 50 * x2 <= 60000
prob += 5 * x1 + 7 * x2 + 8 * x3 <= 10000
prob += 15 * x1 + 25 * x2 + 20 * x3 <= 100000
prob += 5 * x1 <= 50000
prob += 12 * x1 + 8 * x2 + 15 * x3 <= 30000
prob += 200 * x1 + 300 * x2 + 250 * x3 <= 500000
prob += 5 * x3 <= 2000
prob += 5 * x1 + 10 * x2 + 5 * x3 <= 10000
prob += 4 * x1 + 6 * x2 + 8 * x3 <= 30000
prob += 60 * x1 + 90 * x2 + 75 * x3 <= 108000
prob += 45 * x1 + 75 * x2 + 60 * x3 <= 216000

# Solve the problem
prob.solve()

# Create a table to store the results
results_table = pd.DataFrame(columns=['Snack', 'Selling Price (RM)', 'Initial Value', 'Optimized Value', 'Initial Profit (RM)', 'Optimized Profit (RM)'])

# Define the initial values and selling prices
initial_values = {'x1': 800, 'x2': 350, 'x3': 250}
selling_prices = {'x1': 5, 'x2': 7, 'x3': 6}
initial_profits = {snack: initial_values[snack] * selling_prices[snack] for snack in initial_values}

# Add the snack values and profit to the table
initial_value_total = sum(initial_values.values())
optimized_value_total = sum([v.varValue for v in prob.variables()])
for v in prob.variables():
    snack_name = v.name
    initial_value = initial_values[snack_name]
    optimized_value = v.varValue
    selling_price = selling_prices[snack_name]
    initial_profit = initial_profits[snack_name]
    optimized_profit = optimized_value * selling_price
    results_table = results_table.append({'Snack': snack_name, 'Selling Price (RM)': selling_price, 'Initial Value': initial_value, 'Optimized Value': optimized_value,
                                          'Initial Profit (RM)': initial_profit, 'Optimized Profit (RM)': optimized_profit}, ignore_index=True)

# Add the total profit to the table
total_initial_profit = sum(initial_profits.values())
total_optimized_profit = value(prob.objective)
results_table = results_table.append({'Snack': 'Total', 'Selling Price (RM)': '', 'Initial Value': initial_value_total, 'Optimized Value': optimized_value_total,
                                      'Initial Profit (RM)': total_initial_profit, 'Optimized Profit (RM)': total_optimized_profit}, ignore_index=True)

# Round up the values in the table
results_table = results_table.round(decimals=2)

# Set the table format
table_format = 'fancy_grid'

# Print the status of the solution
print("Status:", LpStatus[prob.status])

# Print the results table
print("\nResults:")
print(tabulate(results_table, headers='keys', tablefmt=table_format))

# Print the optimal profit separately
print("\nTotal Initial Profit (RM) =", total_initial_profit)
print("Total Optimized Profit (RM) =", total_optimized_profit)

Status: Optimal

Results:
╒════╤═════════╤══════════════════════╤═════════════════╤═══════════════════╤═══════════════════════╤═════════════════════════╕
│    │ Snack   │ Selling Price (RM)   │   Initial Value │   Optimized Value │   Initial Profit (RM) │   Optimized Profit (RM) │
╞════╪═════════╪══════════════════════╪═════════════════╪═══════════════════╪═══════════════════════╪═════════════════════════╡
│  0 │ x1      │ 5                    │             800 │             962.5 │                  4000 │                  4812.5 │
├────┼─────────┼──────────────────────┼─────────────────┼───────────────────┼───────────────────────┼─────────────────────────┤
│  1 │ x2      │ 7                    │             350 │             350   │                  2450 │                  2450   │
├────┼─────────┼──────────────────────┼─────────────────┼───────────────────┼───────────────────────┼─────────────────────────┤
│  2 │ x3      │ 6                    │             250 │             250   │ 

# <h3><b> F. CONCLUSION

In conclusion, the linear programming model developed for optimizing the production process of snack products has provided valuable insights and solutions. By formulating an objective function to maximize the total profit and imposing various constraints, we were able to determine the optimal production quantities for Cheese Puffs, Potato Chips, and Pretzels.

Based on the results obtained from the optimization process, the optimal production quantities are as follows: 962 units of Cheese Puffs, 350 units of Potato Chips, and 250 units of Pretzels. These quantities were determined by considering the available resources, such as raw materials, machine hours, and labour hours, as well as the selling prices of the snack products.

The maximum profit achieved by the optimal solution is RM8762.50, representing the highest potential earnings from the production process. This optimal solution demonstrates the effective allocation of resources and production quantities to maximize profitability.

Based on the results obtained from the linear programming model, it is worth noting a limitation in the optimization of the production quantities for Potato Chips and Pretzels. The model suggests producing 350 units of Potato Chips and 250 units of Pretzels, which are the lower bounds set for these products. The specified constraints, such as the availability of resources and the production capacity, may restrict the optimal production quantities for Potato Chips and Pretzels. To address this limitation, further analysis and adjustments to the model may be necessary.

Overall, the application of linear programming techniques in optimizing the production process has proven to be a valuable tool for decision-making and maximizing profitability.

# <h3><b> G. REFERENCES

Almeida, F., Barbieri, J. D., Montevechi, J. a. B., Gomes, J. R. B., & Pinho, A. R. (2019). A linear programming optimization model applied to the decision-making process of a Brazilian e-commerce company. Exacta, 17(3), 149–157. https://doi.org/10.5585/exactaep.v17n3.8503



Anggoro, B. S., Rosida, R., Mentari, A. M., Novitasari, C., & Yulista, I. (2019). Profit Optimization Using Simplex Methods on Home Industry Bintang Bakery in Sukarame Bandar Lampung. Journal of Physics, 1155, 012010. https://doi.org/10.1088/1742-6596/1155/1/012010



Uthayageetha, D. (2017). The Maximizing the Profit in Agriculture Field by Simplex Method. International Journal of Advanced Research in Engineering and Technology (IJARET), Volume 8, Issue 6, November-December 2017, pp. 142-146, Available at SSRN https://ssrn.com/abstract=3881003

# <h3><b> H. APPENDIX

In [ ]:
# @title <h5>1. Library</h5>

#Importing Necessary Libraries

# 1. Plotting Libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import cufflinks as cf
%matplotlib inline

# 2. Warning

import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

# 3. Table

from tabulate import tabulate
!pip install tabulate

# 4. Dashboard

!pip install pywedge
import pywedge as pw

# 5. To read pdf

from IPython.display import IFrame

# 6. To convert to HTML

!pip install nbconvert
!pip install pulp
!pip install flask scipy

In [ ]:
import pandas as pd

In [ ]:
# @title <h5>2. To disconnect to Github</h5>

rm -r numericaloptimization # Caution! Only use this when it is absolutely necessary.

In [ ]:
# @title <h5>3. To clone Github</h5>

!git clone https://github.com/Kristian2828/numericaloptimization.git

In [ ]:
# @title <h5>4. Calculator</h5>
from flask import Flask, render_template, request
from scipy.optimize import linprog

app = Flask(__name__)

# Home page
@app.route('/')
def home():
    return render_template('index.html')

# Solver page
@app.route('/solver', methods=['POST'])
def solver():
    # Extracting input values
    x1 = int(request.form['x1'])
    x2 = int(request.form['x2'])
    x3 = int(request.form['x3'])

    # Coefficients of the objective function
    c = [-5, -7, -6]

    # Coefficients of the constraints
    A = [[-150, -200, -250],
         [-50, -5, -7],
         [0, -50, 0],
         [-5, -7, -8],
         [-15, -25, -20],
         [-5, 0, 0],
         [-12, -8, -15],
         [-200, -300, -250],
         [0, 0, -5],
         [-5, -10, -5],
         [-4, -6, -8],
         [-60, -90, -75],
         [-45, -75, -60]]

    # Upper bounds of the constraints
    b = [-350000, -250000, -60000, -10000, -100000, -50000, -30000, -500000, -2000, -10000, -30000, -108000, -216000]

    # Lower bounds of the variables
    x1_bounds = (800, None)
    x2_bounds = (350, None)
    x3_bounds = (250, None)

    # Solving the linear programming problem
    result = linprog(c, A_ub=A, b_ub=b, bounds=[x1_bounds, x2_bounds, x3_bounds], method='simplex')

    if result.success:
        # Extracting the optimized variables
        x1_opt = result.x[0]
        x2_opt = result.x[1]
        x3_opt = result.x[2]

        # Calculating the maximum profit
        max_profit = -result.fun

        return render_template('result.html', x1_opt=x1_opt, x2_opt=x2_opt, x3_opt=x3_opt, max_profit=max_profit)

    return 'Error: Unable to find a solution.'

if __name__ == '__main__':
    app.run(debug=True)
